### PR TITLE
Marginal diagnostics and optimizations

### PR DESCRIPTION
**Pull request by AlexManochis**
This PR contains:

1. two new psrf for 1-dim distribution sampling
2. Optimizations in rounding methods and inner ball computation
3. Birkhoff polytope generator

### CODE REVIEW

File: R-proj/R/gen_birkhoff.R [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  
+  # first column is the vector b
+  b = Mat[,1]
+  Mat = Mat[,-c(1)]
+  
+  P = Hpolytope$new(Mat, b)
+  
+  return(P)
+  
+}
```
Comment (TolisChal) [inline R-proj/R/gen_birkhoff.R]: please add an empty line

File: R-proj/src/rounding.cpp [hunk 2]
```diff
@@ -105,8 +107,8 @@ Rcpp::List rounding (Rcpp::Reference P, Rcpp::Nullable<std::string> method = R_N
                 // low dimensional
                 throw Rcpp::exception("volesti supports rounding for full dimensional polytopes. Maybe call function get_full_dimensional_polytope()");
             }
-            HP.normalize();
             InnerBall = HP.ComputeInnerBall();
+            if (InnerBall.second < 0.0) throw Rcpp::exception("Unable to compute a feasible point.");
             if (method_rcpp.compare(std::string("max_ellipsoid")) == 0) {
                 round_res = max_inscribed_ellipsoid_rounding<MT, VT>(HP, InnerBall);
```
Comment (TolisChal) [inline R-proj/src/rounding.cpp]: Why InnerBall as input? The mentod needs an interior point

File: include/preprocess/min_sampling_covering_ellipsoid_rounding.hpp [hunk 3]
```diff
@@ -109,7 +109,7 @@ std::tuple<MT, VT, NT> min_sampling_covering_ellipsoid_rounding(Polytope &P,
         T = T * L_1.transpose();
 
         P.linear_transformIt(L_1.transpose());
-        P.normalize();
+        //P.normalize();
```
Comment (TolisChal) [inline include/preprocess/min_sampling_covering_ellipsoid_rounding.hpp]: remove if it is not needed

File: include/preprocess/svd_rounding.hpp [hunk 4]
```diff
@@ -153,9 +156,10 @@ std::tuple<MT, VT, NT> svd_rounding(Polytope &P,
             round_it++;
             prev_max_s = max_s;
 
+            P.shift(shift);
             P.linear_transformIt(round_mat);
-            P.normalize();
-            P.ComputeInnerBall();
+            //P.normalize();
```
Comment (TolisChal) [inline include/preprocess/svd_rounding.hpp]: remove it

File: R-proj/R/gen_birkhoff.R (hunk 1 above)
Comment (AlexManochis) [reply R-proj/R/gen_birkhoff.R]: done

File: include/preprocess/min_sampling_covering_ellipsoid_rounding.hpp (hunk 3 above)
Comment (AlexManochis) [reply include/preprocess/min_sampling_covering_ellipsoid_rounding.hpp]: done

File: include/preprocess/svd_rounding.hpp (hunk 4 above)
Comment (AlexManochis) [reply include/preprocess/svd_rounding.hpp]: done

File: R-proj/src/rounding.cpp (hunk 2 above)
Comment (AlexManochis) [reply R-proj/src/rounding.cpp]: Yes I have changed it. Thanks.

File: R-proj/R/gen_birkhoff.R [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#' @examples 
+#' # generate the Birkhoff polytope of order 5
+#' P = gen_birkhoff(5)
+#' @export
+gen_birkhoff <- function(n) {
+  
+  kind_gen = 7
+  m_gen = 0
+  
+  Mat = poly_gen(kind_gen, FALSE, FALSE, n, m_gen)
```
Comment (vissarion) [inline R-proj/R/gen_birkhoff.R:18]: is this working for any value of n? If not comment on it.

File: R-proj/src/psrf_univariate.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+//'
+//' @references \cite{Brooks, S. and Gelman, A.,
+//' \dQuote{General Methods for Monitoring Convergence of Iterative Simulations,} \emph{Journal of Computational and Graphical Statistics,} 1998.}
+//'
+//' @return A vector that contains the values of PSRF for each coordinate
+//'
+//' @export
+// [[Rcpp::export]]
+Rcpp::NumericVector psrf_univariate(Rcpp::NumericMatrix samples,
+                                  Rcpp::Nullable<std::string> method = R_NilValue)
```
Comment (vissarion) [inline R-proj/src/psrf_univariate.cpp]: indentation is wrong

**Review by vissarion — APPROVED** [review]
Thanks for this PR!

File: R-proj/src/psrf_univariate.cpp (hunk 6 above)
Comment (AlexManochis) [reply R-proj/src/psrf_univariate.cpp]: done

**Pull request merged**